[1단계 - 로또 구현] 에드(김진우) 미션 제출합니다.

jinu0328 opened the pull request:
# 1차 리뷰 후 바뀐 주요점

## 1. Factory 제거 
LottosFactory, StatisticsFactory의 경우 객체 생성 방식이 단순하여 객체 생성의 책임이 크지 않다고 판단하여 제거하였습니다. 다만 LottoFactory의 경우 리팩토링하여 유지하는 것을 선택하였는데, 

실제 로또 시스템에서는 자동/수동을 선택하여 구매할 수 있으며, 현재 프로젝트 요구사항에서도

사용자가 구매한 로또 → 자동 생성
지난주 당첨 로또 → 수동 생성
이 적용되므로, 생성 방식을 분리하는 것이 팩토리 패턴을 적용할 이유가 될 수 있다고 생각했습니다.

💡 리팩토링
```java
public class LottoFactory {

    // 수동 로또 생성
    public static Lotto createLotto(List<Integer> numbers) {
        return new Lotto(numbers);
    }

    // 자동 로또 생성
    public static Lotto createRandomLotto() {
        List<Integer> randomNumbers = RandomGenerator.generateNumbers(1, 45, 6); // 유틸 호출
        return new Lotto(randomNumbers);
    }
}
```
이렇게 변경하면,

현재는 객체 생성 방식이 2가지(자동/수동)로 나뉜 것이지만
추후 로또 생성 방식이 더 확장되더라도 쉽게 적용할 수 있는 구조를 가질 수 있다는 점에서 팩토리 패턴이 의미를 가질 수 있을 것 같다고 생각하였습니다!

## 2. 얕은 방어적 복사 사용 
기존 모델 클래스들의 toDto 메서드에서 멤버를 직접 Dto 생성에 활용하여 방어적 복사를 하지 않고있다는 점을 확인하였습니다. 
```java
    // Lotto
    public LottoDto toDto() {
        return new LottoDto(new ArrayList<>(numbers));
    }
    
    // Statistics
    public StatisticsDto toDto() {
        return new StatisticsDto(new HashMap<>(prizeCounts), profitRate);
    }
```
위와 같이 얕은 방어적 복사를 하도록 리팩토링하였습니다!

## 3. 일급컬렉션 Lottos의 불변성 강화
일급컬렉션 멤버에 변경을 가하던 addLotto 메서드를 제거하고 다음과 같이 객체 생성 시점에 초기화 되도록 변경하였습니다.
```java
    public Lottos(List<Lotto> lottos) {
        this.lottos = lottos;
    }
```
### 의문인 점
현재 로또 프로젝트의 경우에는 해당되지 않지만 일급 컬렉션을 적용하는 것이 객체들이 한 번에 초기화되지 않는 경우에도 적절한가? 에 대한 고민이 있습니다. 객체들이 모두 한 번에 초기화되지 않거나, 이후 변경될 가능성이 있는 경우, 일급 컬렉션을 사용하더라도 addLotto()와 같은 추가적인 변경 메서드가 필요해질 것 같습니다. 불변성을 유지 못하더라도 일급컬렉션을 사용한다면 다음과 같은 이점이 있다고 생각합니다. 
- 비즈니스 로직에 종속된 자료구조를 만들 수 있다.
- 여러 객체에 걸친 상태와 행위를 한 곳에서 관리할 수 있다.
- 단순히 컬렉션을 직접 사용하는 것보다 더 뛰어난 가독성을 제공한다.

이러한 점을 고려했을 때, 객체가 한 번에 초기화되지 않는 경우에도 일급 컬렉션을 사용하는 것이 적절할까요? 혹은 다른 대안을 고려하는 것이 좋을까요?

## 4. WinningLotto의 싱글톤 패턴 제거
기존에 한 프로그램 사이클 내에 하나의 WinningLotto 객체만 존재시키기 위해 적용했던 싱글톤 패턴을 제거하였습니다. 이를 통해 테스트도 `@Order` 사용 없이 진행할 수 있도록 리팩토링 하였습니다. 

### 의문인 점
우회적으로 하나의 객체만 존재하게하기 위해 Container나 DI를 활용하는 방식을 시도해보았으나 초기화 시점이 당첨 번호와, 보너스번호를 입력 받은 뒤에 이루어진다는 WinningLotto의 특성 상 쉽게 적용하지 못했던 것 같습니다. 결국 현재는 서비스 레이어에서 new를 통해 객체를 생성하는 상황인데, 객체를 하나만 유지하기 위한 별도의 방법이 있었을 지, 혹은 하나만 유지하게 한다는 생각에 리소스를 더 투자하는 것이 낭비일 지 여쭤보고싶습니다!



---


# 신경쓴 점
1. WinningLotto가 프로그램에서 유일하게 존재하는 당첨 로또라는 의미로 싱글톤 패턴으로 구현하였는데, 우선 테스트할 때 어려움을 겪었습니다. 
5. Getter와 Setter를 최대한 지양하기 위해 Lotto와 WinningLotto 사이에서 양방향 메서드를 사용하거나 DTO를 도입하였습니다. 구현하고보니 getter는 사용하지 않게 되었지만 양방향 메서드를 통한 Lotto와 WinningLotto가 서로를 참조하게 되는 상황입니다. 이에 대해서 그냥 getter를 일부 허용하는 것이 더 나은 방법일지 혹은 다른 개선 방안이 있을 지 궁금합니다.

# Prolog 링크
[Prolog](https://prolog.techcourse.co.kr/studylogs/3835
)
# java-lotto

로또 미션 저장소

# 입력

- [x]  구입 금액을 입력한다
    - [x]  음수면 예외처리
    - [x]  숫자가 아니면 예외처리
    - [x]  1000원으로 안나눠지면 예외처리
- [x]  지난 주 당첨번호를 입력한다
    - [x]  콤마(,)를 구분자로 사용한다
    - [x]  숫자가 6개가 아니면 예외처리
    - [x]  숫자가 아닌 값이 들어오면 예외처리
    - [x]  1~45 범위의 숫자가 아니면 예외처리
    - [x]  중복된 숫자면 예외처리
- [x]  보너스 번호를 입력한다.
    - [x]  앞선 당첨번호랑 중복되면 예외처리
    - [x]  숫자가 아닌 값이 들어오면 예외처리
    - [x]  1~45 범위의 숫자가 아니면 예외처리
- [x] 문자열로 입력 받은 구입 금액을 정수로 변환한다.
- [x] 문자열로 입력 받은 당첨 번호를 정수 리스트로 변환한다.
- [x] 문자열로 입력 받은 보너스 번호를 정수로 변환한다.

---

# 도메인

### 로또 발행

- [x]  구입 금액을 로또 가격으로 나눈만큼 로또를 생성한다
- [x]  로또는 6개의 번호를 가진다
    - [x]  6개의 번호를 중복없이 1~45 랜덤으로 생성한다
        - [x]  생성된 번호를 오름차순 정렬한다

### 당첨 판단

- [x]  당첨 통계를 생성한다
- [x]  당첨 개수에 따른 상금은 다음과 같다

```
3개 일치 (5000원)
4개 일치 (50000원)
5개 일치 (1500000원)
5개 일치, 보너스 볼 일치(30000000원)
6개 일치 (2000000000원)
  ```

- [x]  총 상금을 계산한다
- [x]  수익률을 계산한다

---

# 출력

- [x]  발행된 로또의 개수를 산출하여 출력한다
- [x]  발행된 로또의 번호를 출력한다
- [x]  당첨 통계를 출력한다

jinu0328 committed:
docs(readme): 1차 요구사항 업데이트

jinu0328 committed:
feat(view): 사용자 입력 처리하는 InputView 구현

- 구입 금액 입력 받는 함수 getPurchaseInput 구현

jinu0328 committed:
feat(view): 출력 담당 OutputView 구현

- 구매개수 출력 메서드 구현

jinu0328 committed:
feat(model): Lotto 모델 생성

- 기본 검증 메서드 구현(개수, 범위, 중복)
- 당첨번호 일치, 보너스번호 일치 판단 메서드 구현

jinu0328 committed:
feat(util): 난수 생성 유틸 클래스 구현

jinu0328 committed:
feat(service): 로또 생성 담당 LottoGenerateService 구현

- 생성된 난수 오름차순 정렬하여 Lotto 객체 생성
- 구입 금액 유효성 검사 메서드 구현

jinu0328 committed:
feat(model): Lotto 일급 컬렉션 Lottos 구현

jinu0328 committed:
feat(dto): LottoDto 구현

- Lotto 클래스 내에 toDto 메서드 구현

jinu0328 committed:
feat(view): 로또 번호 출력 메서드 구현

jinu0328 committed:
feat(util): 사용자 입력을 유효값으로 변환하는 InputConverter 클래스 구현

jinu0328 committed:
feat(view): 당첨번호 입력 & 보너스번호 입력 메서드 구현

jinu0328 committed:
feat(model): WinningLotto 모델 클래스 구현

- 싱글턴 패턴 적용
- 보너스 번호 중복 및 범위 유효성 검사 메서드 구현

jinu0328 committed:
feat(model): PrizeTier enum 구현

jinu0328 committed:
feat(model): 당첨 통계를 나타내는 Statistics 모델 구현

jinu0328 committed:
feat(model): Lotto의 PrizeTier 산출 메서드 구현

- Lotto와 WinningLotto 객체의 양방향 메서드로 구현

jinu0328 committed:
feat(service): StatisticsService 구현

jinu0328 committed:
feat(model): PrizeTier 별 개수 & 총 상금 계산 메서드 구현

jinu0328 committed:
feat(dto): 로또 출력을 위한 LottoDto 구현

jinu0328 committed:
feat(dto): Statistics 출력을 위한 StatisticsDto 구현

jinu0328 committed:
feat(model): PrizeTier getter 추가

jinu0328 committed:
chore(model): Statistics의 사용되지 않는 멤버 제거

jinu0328 committed:
feat(view): 당첨 통계 출력 메서드 구현

jinu0328 committed:
docs(readme): 1차 요구사항 구현 완료

jinu0328 committed:
feat(factory): Lotto 생성 담당 클래스 LottoFactory 구현

jinu0328 committed:
feat(factory): Lottos 생성 담당 클래스 LottosFactory 구현

jinu0328 committed:
feat(dto): LottosDto 구현

jinu0328 committed:
feat(service): 로또 추첨 담당 서비스 LottoEvaluationService 구현

jinu0328 committed:
feat(view): ViewFacade 구현

jinu0328 committed:
feat(config): view 및 service 객체 의존성 관리 담당 Container 구현

jinu0328 committed:
feat(controller): LottoController 구현

- 분리된 메서든 간 데이터 전달을 위해 record를 사용

jinu0328 committed:
feat(app): 프로그램 시작점 Application 구현

jinu0328 committed:
chore(view): 출력에 줄바꿈 추가

jinu0328 committed:
fix(view): scanner.next -> scanner.nextLine

jinu0328 committed:
fix(util): 로또 번호 정렬 방식 수정

- toList()를 통한 List는 불변이므로 사전에 정렬 처리

jinu0328 committed:
chore(util): 공백 제거

jinu0328 committed:
fix(service): 구입 금액이 0일 시 오류 방지

jinu0328 committed:
feat(model): 로또 등수 결정 메서드 & dto 변환 메서드 구현

jinu0328 committed:
chore(model): Lotto 모델의 사용되지 않는 멤버 제거

jinu0328 committed:
chore(model): PrizeTier의 사용되지 않는 메서드 제거

jinu0328 committed:
chore(model): WinningLotto의 멤버 이름 변경

- lotto -> basicLotto

jinu0328 committed:
refactor(model): 등수에 해당하는 로또 개수 로직 변경

- 가독성을 위해 for, if문에서 스트림으로 변경

jinu0328 committed:
feat(factory): Statistics 생성 담당 StatisticsFactory 구현

jinu0328 committed:
refactor: 가독성을 위한 메서드 분리

jinu0328 committed:
refactor(error): 에러메시지 Enum 구현 및 적용

jinu0328 committed:
test(model): Lotto 단위 테스트 구현

jinu0328 committed:
test(model): PrizeTier 단위 테스트 구현

jinu0328 committed:
test(model): WinningLotto 단위테스트 구현

jinu0328 committed:
docs(readme): README 수정

yebink commented on src/test/java/model/LottoTest.java:
```diff
@@ -0,0 +1,66 @@
+
+    @ParameterizedTest
+    @MethodSource("provideInvalidLottoNumbersCount")
+    void 번호가_6개가_아니면_예외가_발생한다(List<Integer> numbers) {
```
경계값(5, 7)에 대해 잘 테스트하셨네요! 💯

yebink replied:
다음 단계에서는 `@DisplayName` 에 대해 학습하고 사용해보아요!

yebink commented on src/test/java/model/LottoTest.java:
```diff
@@ -0,0 +1,66 @@
+
+    @ParameterizedTest
+    @MethodSource("provideValidLottoNumbers")
+    void 제대로된_번호가_오면_예외가_발생하지_않는다(List<Integer> numbers) {
```
서로 다른 4가지 경우의 수에 대해 테스트하고 있는데, 작성된 테스트를 읽는 것도 코드 유지보수 비용이랍니다.

4가지 모두 반드시 필요한 테스트일지 고민해보아요. 🙂

yebink commented on src/main/java/factory/LottoFactory.java:
```diff
@@ -0,0 +1,10 @@
+
+public class LottoFactory {
+    public static Lotto createLotto(List<Integer> numbers) {
+        return new Lotto(numbers);
```
`LottoFactory#createLotto` 는 사실상 Lotto 생성자를 한 번 감싼 형태인데요~ 해당 메서드가 필요한 이유가 무엇일까요?

다른 Factory 클래스들도 꼭 필요한 것인지 고민해보아요 😄

jinu0328 replied:
서비스 레이어에서 모델 객체를 직접 new를 통해 생성하기보다 팩토리 클래스를 둠으로써 객체 생성의 책임을 한 곳에서 관리하려 했던 것 같습니다!

엘리의 리뷰를 보고 다시 고민해보니, 현재 Lotto를 비롯한 모델 객체들의 생성 방식이 단순한 편이라 "객체 생성"이라는 책임이 그렇게 무겁지 않음을 깨달았습니다.
즉, 모든 모델 객체에 팩토리 패턴을 적용할 필요는 없을 것 같고, Lottos나 Statistics와 같은 모델에는 팩토리 클래스를 따로 두지 않는 방향이 더 적절할 것 같습니다.

다만, LottoFactory의 경우 다음과 같이 리팩토링 한다면 유지하는 것이 의미가 있을 것 같습니다.
실제 로또 시스템에서는 자동/수동을 선택하여 구매할 수 있으며, 현재 프로젝트 요구사항에서도

사용자가 구매한 로또 → 자동 생성
지난주 당첨 로또 → 수동 생성
이 적용되므로, 생성 방식을 분리하는 것이 팩토리 패턴을 적용할 이유가 될 수 있다고 생각했습니다.

💡 리팩토링 방향

```java
public class LottoFactory {

    // 수동 로또 생성
    public static Lotto createLotto(List<Integer> numbers) {
        return new Lotto(numbers);
    }

    // 자동 로또 생성
    public static Lotto createRandomLotto() {
        List<Integer> randomNumbers = RandomGenerator.generateNumbers(1, 45, 6); // 유틸 호출
        return new Lotto(randomNumbers);
    }
}

```
이렇게 변경하면,

현재는 객체 생성 방식이 2가지(자동/수동)로 나뉜 것이지만
추후 로또 생성 방식이 더 확장되더라도 쉽게 적용할 수 있는 구조를 가질 수 있다는 점에서 팩토리 패턴이 의미를 가질 수 있을 것 같습니다!
피드백 주시면 감사하겠습니다! 🙏😊

yebink commented on src/main/java/model/Lotto.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private void validateNumbersRange(List<Integer> numbers) {
+        for (Integer number : numbers) {
+            if (number < 1 || number > 45) {
+                throw new IllegalArgumentException(ErrorMessage.NUMBER_OUT_OF_RANGE.getMessage());
+            }
+        }
```
`규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.` 규칙에 맞게 고쳐볼까요~?

yebink commented on src/main/java/model/Lottos.java:
```diff
@@ -0,0 +1,42 @@
+
+    public void addLotto(Lotto lotto) {
+        lottos.add(lotto);
+    }
```
일급컬렉션의 중요한 특징 중 하나는 `컬렉션의 불변을 보장한다.` 입니다.
addLotto 메서드는 lottos 에 변경을 가하고 있네요! 😅 

일급컬렉션의 목적에 맞게 코드를 수정해볼까요?
참고 블로그 : https://jojoldu.tistory.com/412

yebink commented:
안녕하세요 에드!

작성해주신 신경쓴 점에 대해 피드백 드려요 😄 

- WinningLotto 싱글톤 관련 
  - 테스트 시 어려움을 겪었음에도 싱글톤으로 작성한 이유는 무엇인가요~? 싱글톤이 아니게 되면 발생하는 문제가 있을까요? 
  - 이와 별개로, 테스트 순서에 따라 성공/실패 여부가 달라진다면 테스트 간 의존이 있다는 것인데요! 만약 신규 요구사항으로 인해 `WinningLottoTest` 에 새로운 테스트가 생겼다고 가정해보아요. 그때마다 `@Order` 를 지정해야겠죠? 실무에서는 하나의 프로젝트에 수 백, 수 천 개의 테스트 코드가 있습니다. 그렇다보니 테스트 간 의존성은 없애는 것이 좋습니다 🙂 
  - 해당 내용은 당장 반영하지 않아도 되니 2단계에서도 충분히 고민해보아요~
-  순환참조 관련
   - `Lotto#rankTier` 에서 WinningLotto 를 참조하고 있군요! 저는 getter를 남발하는 것이 아니라면 사용하는 것이 크게 문제는 아니라고 생각합니다 🙂 의존성 방향에 대한 부분도 아직 고민할 단계가 아니니 걱정하지 마세요! 작성해주신 코드도 충분히 잘하셨습니다. 👏 


`@ParameterizedTest` 를 사용해 테스트를 아주 잘 작성하셨네요! 👏 💯 
`@ParameterizedTest` 는 1개의 테스트로 비슷한 여러 케이스를 묶어서 확인할 수 있다는 장점이 있지만, 코드를 읽기 어렵다는 치명적인 단점이 있습니다. 다음 단계에서는 `@Test` 도 충분히 학습해보시길 바라요 👍

yebink reviewed (changes requested):
안녕하세요 에드~
1단계 리뷰어 엘리입니다 😄 

로또 미션 구현하느라 고생 많으셨어요! 👏 👏 👏 
리뷰 확인해보시고 궁금한 부분이 있으면 편하게 DM 주세요~

jinu0328 reviewed (commented):
첫 리뷰 꼼꼼히 달아주셔서 감사합니다!! 
피드백 주신 부분들 더 깊게 고민해보고 반영하여 다시 리뷰 요청 드리겠습니다 😄

jinu0328 commented:
> * 테스트 시 어려움을 겪었음에도 싱글톤으로 작성한 이유는 무엇인가요~? 싱글톤이 아니게 되면 발생하는 문제가 있을까요?

싱글톤이 아니게 된다면 이미 당첨 로또 객체가 존재하더라도 다른 객체가 추가로 생성될 수 있다는 위험성이 증가하고 한번 생성한 뒤에 해당 당첨 로또 객체를 계속 들고 다니며 코드를 작성해야한다는 단점이 있을 것 같습니다!

싱글톤을 유지하면서 테스트를 구현할 방법으로 `WinningLotto`에 객체 초기화 메서드를 따로 두는 것도 고민해봤지만 테스트만을 위한 메서드가 메인 로직에 포함되는 것은 바람직하지 않다고 생각했습니다.

테스트에서 어려움이 있었던 만큼, 싱글톤을 유지하는 것보다 서비스 레이어에서 1회만 생성 후 주입하는 방식으로 변경하는 것이 더 나은 선택일까요??

jinu0328 committed:
chore: 개행 추가

jinu0328 committed:
refactor(factor): LottoFactory의 로또 생성 메서드를 수동/자동 로또 생성으로 분리

- createCustomLotto는 당첨 로또 생성시, createRandomLotto는 사용자 구매 로또 생성시 사용

jinu0328 committed:
refactor(model): toDto 메서드에서 얕은 방어적 복사를 하도록 수정

jinu0328 committed:
refactor(factory): LottosFactory 제거

- Lottos 객체 생성의 경우 별도의 복잡한 로직이나 여러개의 방식을 제공하지 않으므로 서비스 레이어에서 직접 new로 생성하는 방식으로 변경

jinu0328 committed:
refactor(factory): StatisticsFactory 클래스 제거

- Statistics 객체 생성의 경우 별도의 복잡한 로직이나 여러개의 방식을 제공하지 않으므로 서비스 레이어에서 직접 new로 생성하는 방식으로 변경

jinu0328 committed:
refactor: 일급 컬렉션 Lottos의 불변성 강화

- addLotto 메서드를 제거하고, Lottos 객체를 생성 시점에만 리스트를 설정하도록 변경

jinu0328 committed:
refactor: WinningLotto 싱글톤 제거 및 서비스 레이어에서 직접 생성하도록 변경

- WinningLotto를 싱글톤으로 유지하지 않고, evaluateLottos() 내부에서 직접 생성하도록 수정
- 테스트가 용이하도록 WinningLotto를 명시적으로 생성하는 방식으로 변경

jinu0328 committed:
refactor(test): WinningLotto 생성 테스트 방식 변경

- WinningLotto의 싱글톤 패턴이 제거됨에 따라 Order 제거

jinu0328 committed:
refactor(test, model): Lotto 유효성 테스트 간소화

- @ParameterizedTest 및 @MethodSource 제거하여 단일 테스트로 변경

yebink commented:
1차 리뷰 후 바뀐 주요점 잘 읽었습니다 💯 

- **일급컬렉션 Lottos의 불변성 관련**
  - 좋은 질문이네요! 일급컬렉션의 불변성이 보장되어 있지 않다면, 해당 일급컬렉션을 믿고 사용할 수 있을까요?
  - 객체가 한 번에 초기화되지 않는다면, 아래 코드와 같이 기존 일급컬렉션은 유지하면서 새로운 일급컬렉션을 생성하는 방법도 있어요.
  - 일급컬렉션이 주는 장점도 많지만, 불변을 유지하기 어려운 경우에는 오히려 독이 될 수 있어요. 매번 컬렉션을 일급컬렉션 객체로 변환하는 것도 피로하니까요! 상황에 맞게 사용하는 것이 중요하답니다. 지금은 학습의 목적에 맞게 잘 구현하셨어요. 👏 

 ```java
public class Lottos {
    private final List<Lotto> lottos;

    public Lottos(List<Lotto> lottos) {
        this.lottos = lottos;
    }

    public Lottos addLotto(Lotto lotto) {
        List<Lotto> newLottos = new ArrayList<>(this.lottos);
        newLottos.add(lotto);
        return new Lottos(newLottos); // 새로운 객체 반환
    }
    // 이하생략
}
```

<br/>

- **WinningLotto 싱글톤 관련**
  - 싱글톤 객체는 하나의 애플리케이션 내에서 하나만 생성될 수 있는 객체를 뜻합니다. 어디서든 해당 객체에 접근이 가능하고 (글로벌 접근 허용), 한 번만 생성하기 때문에 생성 비용을 절감할 수 있다는 장점이 있어요.
  - WinningLotto 는 1명의 사용자에 대해 1개여야 하는 것이 맞아요. 그런데 사용자가 늘어난다면 어떨까요? WinningLotto가 싱글턴이어도 문제가 없을까요? 에드의 테스트가 순서에 따라 실패하던 것도 해당 부분과 관련이 있답니다.
  - 지금 하시는 고민도 충분히 의미 있지만 추후 Spring을 학습하실 때 언제 싱글톤이 필요한지 자연스럽게 알게 될 거에요. 🙂 

<br/>

- **LottoFactory (팩토리 패턴) 관련** 
  - [에드의 의견](https://github.com/woowacourse/java-lotto/pull/585#discussion_r1955841778)에 동의해요.
  - 그럼 인터페이스 방식은 어떨까요? 에드가 작성한 팩토리 패턴과 인터페이스 방식에는 어떤 차이가 있을까요? 해당 내용은 2단계에서 더 고민해봐요 😄 

마지막으로, 디자인 패턴을 적용해보려고 많이 노력하신 것 같아요! 👏 
디자인 패턴을 알고 있는 것은 도움이 많이 되지만, 디자인 패턴에 의존하지 않는 연습을 하는 것도 중요하답니다!
다음 미션에서는 관련한 부분도 같이 고민해보시면 좋을 것 같아요.
고생 많으셨습니다 👍 👍 👍

yebink reviewed (approved):
안녕하세요 에드! 

피드백 잘 반영해주셨네요 💯 
추가 코멘트 남겼으니 확인해주세요.
2단계 미션에서 뵐게요 😄
